File: py_src/taskito/resources/runtime.py
"""ResourceRuntime — initializes, resolves, and tears down worker resources."""

from __future__ import annotations

import logging
from collections.abc import Callable
from typing import Any

from taskito.async_support.helpers import run_maybe_async
from taskito.exceptions import (
    ResourceInitError,
    ResourceNotFoundError,
    ResourceUnavailableError,
)
from taskito.resources.definition import ResourceDefinition, ResourceScope
from taskito.resources.graph import topological_sort

logger = logging.getLogger("taskito.resources")


class ResourceRuntime:
    """Manages the lifecycle of scoped resources.

    Worker-scoped resources are initialized eagerly in dependency order.
    Task-scoped resources get a pool. Thread-scoped resources use thread-local storage.
    Request-scoped resources are created fresh per task.
    """

    def __init__(self, definitions: dict[str, ResourceDefinition]) -> None:
        self._definitions = dict(definitions)
        self._instances: dict[str, Any] = {}
        self._init_order: list[str] = []
        self._unhealthy: set[str] = set()
        self._recreation_count: dict[str, int] = {}
        self._init_duration: dict[str, float] = {}
        self._pools: dict[str, Any] = {}  # ResourcePool instances
        self._thread_locals: dict[str, Any] = {}  # ThreadLocalStore instances

    def initialize(self) -> None:
        """Create all resources in topological (dependency-first) order."""
        import time

        from taskito.resources.frozen import FrozenResource
        from taskito.resources.pool import PoolConfig, ResourcePool
        from taskito.resources.thread_local import ThreadLocalStore

        self._init_order = topological_sort(self._definitions)
        for name in self._init_order:
            defn = self._definitions[name]
            start = time.monotonic()

            if defn.scope == ResourceScope.WORKER:
                self._create_resource(name)
                if defn.frozen and name in self._instances:
                    self._instances[name] = FrozenResource(self._instances[name], name)
            elif defn.scope == ResourceScope.TASK:
                pool_size = defn.pool_size or 4
                dep_kwargs = {dep: self._instances[dep] for dep in defn.depends_on}
                pool = ResourcePool(
                    name=name,
                    factory=defn.factory,
                    teardown=defn.teardown,
                    config=PoolConfig(
                        pool_size=pool_size,
                        pool_min=defn.pool_min,
                        acquire_timeout=defn.acquire_timeout,
                        max_lifetime=defn.max_lifetime,
                        idle_timeout=defn.idle_timeout,
                    ),
                    dep_kwargs=dep_kwargs,
                )
                if defn.pool_min > 0:
                    pool.prewarm()
                self._pools[name] = pool
            elif defn.scope == ResourceScope.THREAD:
                dep_kwargs = {dep: self._instances[dep] for dep in defn.depends_on}
                store = ThreadLocalStore(
                    name=name,
                    factory=defn.factory,
                    teardown=defn.teardown,
                    dep_kwargs=dep_kwargs,
                )
                self._thread_locals[name] = store
            elif defn.scope == ResourceScope.REQUEST:
                pass  # created fresh in acquire_for_task

            self._init_duration[name] = time.monotonic() - start

    def _create_resource(self, name: str) -> None:
        """Invoke a resource factory, injecting its declared dependencies."""
        defn = self._definitions[name]
        dep_kwargs = {dep: self._instances[dep] for dep in defn.depends_on}
        try:
            result = run_maybe_async(defn.factory(**dep_kwargs))
            self._instances[name] = result
            self._unhealthy.discard(name)
        except Exception as exc:
            raise ResourceInitError(f"Failed to initialize resource '{name}': {exc}") from exc

    def resolve(self, name: str) -> Any:
        """Return a live resource instance by name (worker scope only).

        For task/request scopes, use acquire_for_task() instead.

        Raises:
            ResourceNotFoundError: If the name was never registered.
            ResourceUnavailableError: If the resource is permanently unhealthy.
        """
        if name not in self._definitions and name not in self._instances:
            raise ResourceNotFoundError(f"Resource '{name}' is not registered")
        if name in self._unhealthy:
            raise ResourceUnavailableError(f"Resource '{name}' is permanently unhealthy")

        # Worker scope — direct return
        if name in self._instances:
            return self._instances[name]

        # Thread scope — thread-local access
        if name in self._thread_locals:
            return self._thread_locals[name].get_or_create()

        # Task/request scope shouldn't use resolve() directly
        if name in self._pools:
            raise ResourceUnavailableError(
                f"Resource '{name}' is task-scoped — use acquire_for_task() instead"
            )

        raise ResourceNotFoundError(f"Resource '{name}' is not initialized")

    def acquire_for_task(self, name: str) -> tuple[Any, Callable[[], None] | None]:
        """Acquire a resource for a single task execution.

        Returns:
            (instance, release_callback) where release_callback is None
            for worker/thread scopes and must be called after task completion
            for task/request scopes.
        """
        if name not in self._definitions and name not in self._instances:
            raise ResourceNotFoundError(f"Resource '{name}' is not registered")
        if name in self._unhealthy:
            raise ResourceUnavailableError(f"Resource '{name}' is permanently unhealthy")

        defn = self._definitions.get(name)
        scope = defn.scope if defn else ResourceScope.WORKER

        if scope == ResourceScope.WORKER:
            return self._instances[name], None

        if scope == ResourceScope.THREAD:
            return self._thread_locals[name].get_or_create(), None

        if scope == ResourceScope.TASK:
            pool = self._pools[name]
            instance = pool.acquire()

            def release() -> None:
                pool.release(instance)

            return instance, release

        if scope == ResourceScope.REQUEST:
            deps = defn.depends_on if defn else []
            dep_kwargs = {dep: self._instances.get(dep) for dep in deps}
            instance = defn.factory(**dep_kwargs) if defn else None
            instance = run_maybe_async(instance)

            def teardown_request() -> None:
                if defn and defn.teardown is not None and instance is not None:
                    try:
                        defn.teardown(instance)
                    except Exception:
                        logger.exception("Error tearing down request resource '%s'", name)

            return instance, teardown_request

        return self._instances.get(name), None

    def recreate(self, name: str) -> bool:
        """Attempt to recreate a single resource. Returns True on success."""
        try:
            old = self._instances.get(name)
            defn = self._definitions[name]
            if old is not None and defn.teardown is not None:
                run_maybe_async(defn.teardown(old))
            self._create_resource(name)
            self._recreation_count[name] = self._recreation_count.get(name, 0) + 1
            return True
        except ResourceInitError:
            return False

    def reload(self, names: list[str] | None = None) -> dict[str, bool]:
        """Reload reloadable resources. Returns {name: success} dict."""
        results: dict[str, bool] = {}
        targets = names or [n for n, d in self._definitions.items() if d.reloadable]
        # Reload in dependency order
        for name in self._init_order:
            if name not in targets:
                continue
            defn = self._definitions.get(name)
            if defn is None or (names is None and not defn.reloadable):
                continue
            results[name] = self.recreate(name)
        return results

    def mark_unhealthy(self, name: str) -> None:
        """Mark a resource as permanently unhealthy."""
        self._unhealthy.add(name)
        logger.error("Resource '%s' marked permanently unhealthy", name)

    def teardown(self) -> None:
        """Tear down all resources in reverse initialization order."""
        # Shutdown pools first
        for pool in self._pools.values():
            pool.shutdown()
        self._pools.clear()

        # Teardown thread-local stores
        for store in self._thread_locals.values():
            store.teardown_all()
        self._thread_locals.clear()

        # Teardown worker-scoped resources in reverse order
        for name in reversed(self._init_order):
            defn = self._definitions.get(name)
            instance = self._instances.pop(name, None)
            if defn is not None and defn.teardown is not None and instance is not None:
                try:
                    run_maybe_async(defn.teardown(instance))
                except Exception:
                    logger.exception("Error tearing down resource '%s'", name)
        self._init_order.clear()

    def status(self) -> list[dict[str, Any]]:
        """Return a summary of each resource's current state."""
        result: list[dict[str, Any]] = []
        for name in self._init_order:
            defn = self._definitions.get(name)
            if name in self._unhealthy:
                health = "unhealthy"
            elif name in self._instances or name in self._pools or name in self._thread_locals:
                health = "healthy"
            else:
                health = "unknown"
            entry: dict[str, Any] = {
                "name": name,
                "scope": defn.scope.value if defn else "worker",
                "health": health,
                "init_duration_ms": round(self._init_duration.get(name, 0) * 1000, 2),
                "recreations": self._recreation_count.get(name, 0),
                "depends_on": defn.depends_on if defn else [],
            }
            # Add pool stats for task-scoped resources
            if name in self._pools:
                entry["pool"] = self._pools[name].stats()
            result.append(entry)
        return result

    @classmethod
    def from_test_overrides(cls, instances: dict[str, Any]) -> ResourceRuntime:
        """Create a runtime pre-populated with mock/test instances.

        No factories are called — the provided instances are used directly.
        """
        rt = cls(definitions={})
        rt._instances = dict(instances)
        rt._init_order = list(instances.keys())
        return rt

File: sdks/node/src/resources/runtime.ts
import {
  ResourceError,
  ResourceNotFoundError,
  ResourceScopeError,
  ResourceUnavailableError,
} from "../errors";
import { createLogger } from "../utils";
import { HealthChecker } from "./health";
import { ResourcePool } from "./pool";
import type {
  ResourceContext,
  ResourceDefinition,
  ResourceMetrics,
  ResourceResolver,
  ResourceScope,
} from "./types";
import { RESOURCE_SCOPES } from "./types";

const log = createLogger("resources");

/** Shared empty ancestry for a top-level resolve. */
const EMPTY_CHAIN: ReadonlySet<string> = new Set();

/** A disposal thunk plus the resource name, for error context. */
interface Teardown {
  name: string;
  run: () => void | Promise<void>;
  /** The build this thunk disposes, so recreation can retire exactly it. */
  built?: Promise<unknown>;
}

/** Per-resource health, as advertised in the worker heartbeat. */
export type HealthState = "healthy" | "unhealthy";

/** Per-invocation task scope: caches task-scoped resources and disposes them. */
export interface TaskScope {
  /** Resolve a resource for this invocation (task cache first, then worker). */
  readonly resolver: ResourceResolver;
  /** Dispose the task-scoped resources built during this invocation, LIFO. */
  teardown(): Promise<void>;
}

/**
 * Registry + lifecycle for injectable resources. Worker-scoped values are built
 * once and shared; task-scoped values are built per invocation; pooled values
 * are checked out of a bounded pool per invocation and returned afterwards. The
 * worker wires this into task execution; tasks reach values via `useResource`
 * or declarative `inject`.
 *
 * Every resolver always returns a promise — guard and factory failures surface as
 * rejections, never synchronous throws — so a failed build is awaited and retried,
 * not cached.
 */
export class ResourceRuntime {
  private readonly defs = new Map<string, ResourceDefinition>();
  private readonly workerCache = new Map<string, Promise<unknown>>();
  private readonly workerTeardown: Teardown[] = [];
  private readonly counters = new Map<string, { created: number; disposed: number }>();
  /** Checkout pools behind pooled-scope resources, built lazily per resource. */
  private readonly pools = new Map<string, ResourcePool>();
  /** Active worker leases sharing this runtime; teardown disposes only at zero. */
  private workerLeases = 0;
  /** Resources marked permanently unhealthy — resolving them rejects for good. */
  private readonly unhealthy = new Set<string>();
  /** One checker per runtime, shared by every lease — never one per worker. */
  private healthChecker: HealthChecker | undefined;

  /** Register (or replace) a resource definition. */
  register<T>(name: string, definition: ResourceDefinition<T>): void {
    if (!(RESOURCE_SCOPES as readonly string[]).includes(definition.scope)) {
      throw new ResourceError(
        `resource "${name}": unknown scope "${definition.scope}" — expected one of ${RESOURCE_SCOPES.join(", ")}`,
      );
    }
    this.defs.set(name, definition as ResourceDefinition);
    // A replacement definition starts with a clean bill of health.
    this.unhealthy.delete(name);
    // Drop any built worker instance so "replace" takes effect; the old
    // instance is still disposed by its queued teardown.
    this.workerCache.delete(name);
    // Likewise retire the old pool; its idle instances are disposed best-effort.
    const stalePool = this.pools.get(name);
    if (stalePool) {
      this.pools.delete(name);
      void stalePool.shutdown();
    }
  }

  /** True when nothing is registered — lets the worker skip resource wiring. */
  get isEmpty(): boolean {
    return this.defs.size === 0;
  }

  /** Snapshot of per-resource lifecycle counters (created / disposed / active). */
  metrics(): ResourceMetrics {
    const out: ResourceMetrics = {};
    for (const [name, c] of this.counters) {
      out[name] = { created: c.created, disposed: c.disposed, active: c.created - c.disposed };
    }
    return out;
  }

  /** Get-or-create the mutable counter for a resource. */
  private counter(name: string): { created: number; disposed: number } {
    let c = this.counters.get(name);
    if (!c) {
      c = { created: 0, disposed: 0 };
      this.counters.set(name, c);
    }
    return c;
  }

  /** Build a worker-scoped resource once, memoizing the promise (dedups concurrent init). */
  private resolveWorker(name: string): Promise<unknown> {
    if (this.unhealthy.has(name)) {
      return Promise.reject(
        new ResourceUnavailableError(`Resource "${name}" is permanently unhealthy`),
      );
    }
    const cached = this.workerCache.get(name);
    if (cached) {
      return cached;
    }
    const def = this.defs.get(name);
    if (!def) {
      return Promise.reject(unregistered(name));
    }
    if (def.scope !== "worker") {
      return Promise.reject(new ResourceScopeError(name, def.scope));
    }
    const ctx: ResourceContext = {
      scope: "worker",
      use: <T>(dep: string) => this.resolveWorker(dep) as Promise<T>,
    };
    const counter = this.counter(name);
    counter.created += 1;
    // Register the pending promise BEFORE the factory body runs (deferred a
    // microtask) so a resource that resolves a dependency on itself reuses this
    // promise instead of recursing into resolveWorker and overflowing the stack.
    const built = Promise.resolve()
      .then(() => startFactory(def, ctx))
      .catch((error) => {
        counter.created -= 1; // a failed build is not a live resource
        this.workerCache.delete(name); // failed build is retryable, not cached
        throw error;
      });
    this.workerCache.set(name, built);
    this.trackResource(this.workerTeardown, name, def, built);
    return built;
  }

  /**
   * Resolve a dependency that must be worker-scoped. Pooled instances outlive
   * any single task, so their factories may only depend on worker singletons.
   */
  private resolveWorkerDependency(name: string, context: ResourceScope): Promise<unknown> {
    const def = this.defs.get(name);
    if (!def) {
      return Promise.reject(unregistered(name));
    }
    if (def.scope !== "worker") {
      return Promise.reject(new ResourceScopeError(name, def.scope, context));
    }
    return this.resolveWorker(name);
  }

  /** Get-or-create the checkout pool behind a pooled-scope resource. */
  private poolFor(name: string, def: ResourceDefinition): ResourcePool {
    let pool = this.pools.get(name);
    if (!pool) {
      const ctx: ResourceContext = {
        scope: "pooled",
        use: <T>(dep: string) => this.resolveWorkerDependency(dep, "pooled") as Promise<T>,
      };
      const counter = this.counter(name);
      pool = new ResourcePool(name, () => startFactory(def, ctx), def.dispose, def.pool ?? {}, {
        onCreated: () => {
          counter.created += 1;
        },
        onDisposed: () => {
          counter.disposed += 1;
        },
      });
      this.pools.set(name, pool);
    }
    return pool;
  }

  /** Begin a per-invocation task scope. */
  createTaskScope(): TaskScope {
    const taskCache = new Map<string, Promise<unknown>>();
    const taskTeardown: Teardown[] = [];

    const resolve = (
      name: string,
      building: ReadonlySet<string> = EMPTY_CHAIN,
    ): Promise<unknown> => {
      const def = this.defs.get(name);
      if (!def) {
        return Promise.reject(unregistered(name));
      }
      if (def.scope === "worker") {
        return this.resolveWorker(name); // shared singleton, even when first reached here
      }
      if (def.scope === "pooled") {
        const pending = taskCache.get(name);
        if (pending) {
          return pending; // one checkout per task, shared by every resolve
        }
        const pool = this.poolFor(name, def);
        // Cache the pending checkout BEFORE awaiting (see the task branch below)
        // so concurrent resolves within the task share a single checkout.
        const checkout = pool.acquire().catch((error) => {
          taskCache.delete(name); // failed checkout is retryable, not cached
          throw error;
        });
        taskCache.set(name, checkout);
        taskTeardown.push({
          name,
          run: async () => {
            let value: unknown;
            try {
              value = await checkout;
            } catch {
              return; // checkout failed — nothing to return to the pool
            }
            await pool.release(value); // return, don't dispose: the instance stays pooled
          },
        });
        return checkout;
      }
      if (def.scope === "request") {
        // Fresh on every resolve, never cached: N `useResource()` calls inside one
        // task yield N instances, each disposed with the task (LIFO, like the rest).
        // Being uncached is also why a cycle cannot resolve itself the way a task
        // resource does — there is no pending promise to hand back, so track the
        // chain and reject instead of recursing forever.
        if (building.has(name)) {
          return Promise.reject(
            new ResourceError(
              `resource "${name}": request-scope dependency cycle (${[...building, name].join(" -> ")})`,
            ),
          );
        }
        const chain = new Set(building).add(name);
        const ctx: ResourceContext = {
          scope: "request",
          use: <T>(dep: string) => resolve(dep, chain) as Promise<T>,
        };
        const counter = this.counter(name);
        counter.created += 1;
        const built = Promise.resolve()
          .then(() => startFactory(def, ctx))
          .catch((error) => {
            counter.created -= 1; // a failed build is not a live resource
            throw error;
          });
        this.trackResource(taskTeardown, name, def, built);
        return built;
      }
      const cached = taskCache.get(name);
      if (cached) {
        return cached;
      }
      const ctx: ResourceContext = {
        scope: "task",
        use: <T>(dep: string) => resolve(dep, building) as Promise<T>,
      };
      const counter = this.counter(name);
      counter.created += 1;
      // Cache before the factory runs (see resolveWorker) so a self-referential
      // task resource reuses this promise instead of recursing.
      const built = Promise.resolve()
        .then(() => startFactory(def, ctx))
        .catch((error) => {
          counter.created -= 1; // a failed build is not a live resource
          taskCache.delete(name); // failed build is retryable, not cached
          throw error;
        });
      taskCache.set(name, built);
      this.trackResource(taskTeardown, name, def, built);
      return built;
    };

    const resolver: ResourceResolver = (name) => resolve(name);
    return { resolver, teardown: () => runTeardown(taskTeardown) };
  }

  /** Register a worker that shares this runtime's worker-scoped resources. */
  acquireWorker(): void {
    this.workerLeases += 1;
    // Start the shared checker on the first lease only — concurrent workers on
    // one runtime must not race recreate() on the same resource.
    if (this.workerLeases === 1) {
      this.healthChecker = new HealthChecker(this);
      this.healthChecker.start();
    }
    // Best-effort prewarm of pooled resources that asked for warm instances.
    // `prewarm` never rejects — build failures are logged inside the pool.
    for (const [name, def] of this.defs) {
      if (def.scope === "pooled" && (def.pool?.poolMin ?? 0) > 0) {
        void this.poolFor(name, def).prewarm();
      }
    }
  }

  /**
   * Release one worker lease; dispose worker-scoped resources (LIFO) and reset
   * the caches only once the last sharing worker has stopped. This keeps a
   * second `runWorker()` on the same queue from tearing down resources the
   * first worker is still using.
   */
  async teardownWorker(): Promise<void> {
    if (this.workerLeases > 0) {
      this.workerLeases -= 1;
    }
    if (this.workerLeases > 0) {
      return;
    }
    // Drain the checker first: once stop() resolves, no in-flight tick can
    // recreate a resource while the caches below are torn down.
    const checker = this.healthChecker;
    this.healthChecker = undefined;
    await checker?.stop();
    const pending = this.workerTeardown.splice(0);
    this.workerCache.clear();
    // "Permanently unhealthy" is scoped to a worker run — the next worker
    // starts from scratch and may succeed where this one gave up.
    this.unhealthy.clear();
    // Shut pools down first — pooled instances may depend on worker resources.
    // `shutdown` never throws; per-instance dispose failures are logged.
    const pools = [...this.pools.values()];
    this.pools.clear();
    for (const pool of pools) {
      await pool.shutdown();
    }
    await runTeardown(pending);
  }

  // ── Health ────────────────────────────────────────────────────────────────

  /** Registered resource names, sorted — advertised on worker registration. */
  get names(): string[] {
    return [...this.defs.keys()].sort();
  }

  /** Definitions that asked for periodic health checks (interval > 0). */
  healthChecked(): ReadonlyMap<string, ResourceDefinition> {
    const out = new Map<string, ResourceDefinition>();
    for (const [name, def] of this.defs) {
      if (def.healthCheck && (def.healthCheckIntervalMs ?? 0) > 0) {
        out.set(name, def);
      }
    }
    return out;
  }

  /** Whether `name` was marked permanently unhealthy. */
  isUnhealthy(name: string): boolean {
    return this.unhealthy.has(name);
  }

  /** The built (possibly pending) worker instance, or undefined before first build. */
  builtWorkerInstance(name: string): Promise<unknown> | undefined {
    return this.workerCache.get(name);
  }

  /**
   * Mark a resource permanently unhealthy: every later resolve rejects with
   * {@link ResourceUnavailableError}. Terminal — there is no auto-recovery.
   */
  markUnhealthy(name: string): void {
    this.unhealthy.add(name);
    log.error(`resource "${name}" marked permanently unhealthy`);
  }

  /**
   * Dispose the current worker instance (best effort) and build a fresh one.
   * Returns false when the factory fails — the caller decides whether that
   * exhausts the recreation budget.
   */
  async recreate(name: string): Promise<boolean> {
    const def = this.defs.get(name);
    if (!def) {
      return false;
    }
    await this.disposeReplacedInstance(name, def);
    try {
      await this.resolveWorker(name);
      return true;
    } catch (error) {
      log.warn(() => `recreating resource "${name}" failed`, error);
      return false;
    }
  }

  /**
   * Per-resource health for the worker heartbeat, or undefined when nothing is
   * registered (so the heartbeat clears the column instead of writing `{}`).
   */
  healthSnapshot(): Record<string, HealthState> | undefined {
    if (this.defs.size === 0) {
      return undefined;
    }
    const out: Record<string, HealthState> = {};
    for (const name of this.defs.keys()) {
      out[name] = this.unhealthy.has(name) ? "unhealthy" : "healthy";
    }
    return out;
  }

  /**
   * Retire the currently-built worker instance ahead of recreation: drop it
   * from the cache, cancel its queued teardown (recreation disposes it now),
   * and run its `dispose` best-effort.
   */
  private async disposeReplacedInstance(name: string, def: ResourceDefinition): Promise<void> {
    const old = this.workerCache.get(name);
    if (!old) {
      return;
    }
    this.workerCache.delete(name);
    const staleIndex = this.workerTeardown.findIndex((entry) => entry.built === old);
    if (staleIndex !== -1) {
      this.workerTeardown.splice(staleIndex, 1);
    }
    try {
      const value = await old;
      await def.dispose?.(value);
      this.counter(name).disposed += 1;
    } catch (error) {
      log.debug(() => `disposing resource "${name}" before recreation failed`, error);
    }
  }

  /**
   * Queue a teardown thunk for a resource: run its `dispose` (if any) and record
   * the disposal. A build that failed is skipped — there's nothing to tear down.
   */
  private trackResource(
    stack: Teardown[],
    name: string,
    def: ResourceDefinition,
    built: Promise<unknown>,
  ): void {
    const { dispose } = def;
    const counter = this.counter(name);
    stack.push({
      name,
      built,
      run: async () => {
        let value: unknown;
        try {
          value = await built;
        } catch {
          return; // build failed (and was already un-counted) — nothing to dispose
        }
        if (dispose) {
          await dispose(value);
        }
        counter.disposed += 1;
      },
    });
  }
}

/** Run a factory, converting a synchronous throw into a rejected promise. */
function startFactory(def: ResourceDefinition, ctx: ResourceContext): Promise<unknown> {
  return new Promise((resolve) => resolve(def.factory(ctx)));
}

function unregistered(name: string): ResourceNotFoundError {
  return new ResourceNotFoundError(name);
}

/** Run disposal thunks in reverse (LIFO) order; surface the first error after all run. */
async function runTeardown(stack: Teardown[]): Promise<void> {
  let firstError: unknown;
  for (let i = stack.length - 1; i >= 0; i--) {
    try {
      await stack[i]?.run();
    } catch (error) {
      firstError ??= error;
    }
  }
  if (firstError !== undefined) {
    throw firstError;
  }
}
